test/testsymboldatabase.cpp

```diff
@@ -1281,6 +1282,167 @@ class TestSymbolDatabase: public TestFixture {
+                    ASSERT(nullptr == t->variable());
+                }
+            }
+        }
+
+        //Test a more complicated case where foo has been declared as a function so it's definitely a function call
+        {
+            givenACodeSampleToTokenize var("int* a[5] = { new int, new int, new int, new int, new int };\n"
+                "void foo(int x){}\n"
+                "int main() {\n"
+                "foo(*a[4]);\n" //From context this is definitely a function call
+                "}\n");
+            const Function* f = Token::findmatch(var.tokens(), "foo")->function();
+            const Variable* v = Token::findmatch(var.tokens(), "a")->variable();
+            for (auto t : var.tokens()->toend())
+            {
+                if (t->str() == "foo") { //foo is always a function
+                    ASSERT(f == t->function());
+                }
+
+                if (t->str() == "a") { //a is a variable
+                    ASSERT(v == t->variable());
+                }
+                else if (t->str() != "x") { //x is too but everything else is not
+                    ASSERT(nullptr == t->variable());
+                }
+            }
+        }
+
+        //Test a case where foo has been declared as a function in an awkward scope so it's definitely a function call
+        //{
+        //    givenACodeSampleToTokenize var("int* a[5] = { new int, new int, new int, new int, new int };\n"
+        //                                   "namespace X::Y {\n"
+        //                                   "template <typename T, int N1>\n"
+        //                                   "struct bar {\n"
+        //                                   "  template<int N2> static void foo(int x){}\n"
+        //                                   "};\n"
+        //                                   "}\n"
+        //                                   "int main() {\n"
+        //                                   "X::Y::bar<int, 1>::foo<4>(*a[4]);\n" //From context this is definitely a function call
+        //                                   "}\n");
+        //    const Function* f = Token::findmatch(var.tokens(), "foo")->function();
+        //    const Variable* v = Token::findmatch(var.tokens(), "a")->variable();
+        //    for (auto t : var.tokens()->toend())
+        //    {
+        //        if (t->str() == "foo") { //foo is always a function
+        //            ASSERT(f == t->function());
+        //        }
+        //
+        //        if (t->str() == "a") { //a is a variable
+        //            ASSERT(v == t->variable());
+        //        }
+        //        else if (t->str() != "x") { //x is too but everything else is not
+        //            ASSERT(nullptr == t->variable());
+        //        }
+        //    }
+        //}
+
+        //Test the simple case where foo has been declared as a type so it's definitely a variable declaration
```

Comment:
This test case is technically valid and I think cppcheck's interpretation of this code is technically incorrect but honestly it's not a huge concern. It's some weird code that looks almost deliberately obfuscated.